examples/sleuth.py

```diff
@@ -0,0 +1,299 @@
+                typ=str,
+                required=True
+            )
+        }
+    )
+})
+
+
+def get_system_message() -> str:
+    """Get the system message."""
+    return (
+        "You are a business analyst for Tectonic Tech, Inc.  You have the ability to run sqlite queries "
+        "against the company's databses to collect information for the user.\n\n "
+        "Answer the user's questions as best as you can.  Only use the functions you have been "
+        "provided with.\n\n"
+        "Before running a query check to see if you already have part or all of the answer from "
+        "your interaction history!\n\n"
+    )
+
+
+def get_schema_description(databases) -> str:
+    """
+    Construct a description of the DB schema for the LLM by retrieving the
+    CREATE commands used to create the tables.
+    """
+    schema_description = f"There are {len(databases)} databases  Here are their names and schemas:\n"
+    for i,db in enumerate(databases.keys()):
+        schema_description += f'{i+1}. {db}\n'
+        cur = databases[db]['connection'].cursor()
+        tables = cur.execute("select name from sqlite_master where type='table';").fetchall()
+        tables = [col[0] for col in tables]
+        for table in tables:
+            schema_description += f'Table `{table}:`\n'
+            schema_description += str(get_table_schema(databases[db]['connection'], table)) + '\n'
+        schema_description += '\n'
+
+    return schema_description
+
+
+def format_message(message):
+    """Convert a message into plain text"""
+    text = f'{message.role.capitalize()}:\n'
+    text += message.content
+    text += f'\n\n_<name={message.name}, function_call={message.function_call}>_\n'
+    return text
+
+
+def concat_chat_history(chat_history: list[tuple[str,str]]) -> str:
+    """Stringify a chat history."""
+    message_list = [msg[0] if msg[0] else msg[1] for msg in chat_history]
+    return '\n\n'.join(message_list)
+
+
+def main():
+    """Setup and run gradio app."""
+
+    # Setup
+    key = openai.load_api_key(project.OPENAI_API_KEY_FILE_PATH)
+    openai.set_api_key(key)
+
+    # Get a model
+    model = openai.start_chat('gpt-4-0613')
+
+    # Connect to the DB and get the table names
+    log.debug('Getting the DB information.')
+    for db in DATABASES:
+        DATABASES[db]['connection'] = sqlite3.connect(DATABASES[db]['file'])
+    schema_description = get_schema_description(DATABASES)
+
+    for db in DATABASES:
+        DATABASES[db]['connection'].close()
+        DATABASES[db]['connection'] = None
+
+    # Setup the agent
+    system_message = get_system_message()
+    system_message += schema_description
+    print(system_message)
+
+    def new_question_wrapper():
+        return ''
+
+    def agent_wrapper(question, status, chat_history):
+
+        # Build the functions that the agent can use
+        for db in DATABASES:
+            DATABASES[db]['connection'] = sqlite3.connect(DATABASES[db]['file'])
```

Comment:
I know there are some issues with accessing SQLite connections across threads. I'm sure what you have here works, but for the sake of clarity I would suggested making a local copy of the `DATABASES` dict or better yet make a local `db_ connections` dict rather than assigning these connections into a global data structure.